zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -323,39 +361,54 @@ impl Chain {
             orchard_nullifiers: self.orchard_nullifiers.clone(),
             partial_cumulative_work: self.partial_cumulative_work,
             history_tree,
+            chain_value_pools: self.chain_value_pools,
         }
     }
 }
 
-/// Helper trait to organize inverse operations done on the `Chain` type. Used to
-/// overload the `update_chain_state_with` and `revert_chain_state_with` methods
-/// based on the type of the argument.
+/// The revert position being performed on a chain.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+enum RevertPosition {
+    /// The chain root is being reverted via [`pop_root`],
+    /// when a block is finalized.
+    Root,
+
+    /// The chain tip is being reverted via [`pop_tip`],
+    /// when a chain is forked.
+    Tip,
+}
```

Comment:
All the changes related to the position then reversing were a surprise to me but they make sense and seems to be well implemented. Very nice work.